### PR TITLE
feat(pwa): track online status [LIBS-315]

### PR DESCRIPTION
**Pull request by KaiVandivier**
Adds the service worker, offline interface, and app adapter foundation for tracking the DHIS2 connection status alongside incidental requests.

Part of [LIBS-315](https://dhis2.atlassian.net/browse/LIBS-315)

### The spec
* We want to detect connection to the DHIS2 server.
* We want to minimize the network and server resources used to do so
* The service worker watches app network traffic and uses the success or failure of responses to determine connectivity status, then send messages to the client to broadcast the updates
    * Only requests to the DHIS2 server (based on the Base URL) are used to determine the connection status
* If there is no app network traffic to the DHIS2 server, the client will send pings
    * The service worker is configured to only handle pings over the network, i.e. not to cache ping requests/responses, so the result observed on the client will reflect the server connection status
    * If the connection status detected by the ping is the same as the previous status, then the time to the next ping will increase. This will continue to create an exponential back-off, which will max at out at a value, for example 5 minutes.
    * If the connection status has changed (either by ping or by app network traffic), the exponential back-off will be reset and the timeout to the next ping will be reduced back to the initial timeout.
    * If the app loses focus, a “standby” procedure is followed to stop pinging when it’s unnecessary
        * Nothing happens initially. If the user refocuses the app before the next ping is scheduled to be sent, the behavior will not change.
        * If the next-scheduled ping would take place while the app is defocused, instead the hook enters a “standby” mode.
        * When the app is refocused, if the hook is in “standby”, a ping will be sent immediately to detect the current connection status, following other rules for back-off
        * If the app is opened while defocused initially, it will also start running in standby and will not ping until refocused.
    * If the app detects an ‘offline’ event from the browser, it will ping immediately to check the connection status, since that should be a low-cost action
    * If the app detects an ‘offline’ event from the browser and is defocused, a special thing happens: it enters standby to send a ping imediately when the app is refocused again
        * If the app is refocused before the next regularly-scheduled ping, the ping timeout will be restarted but not incremented.
        * Otherwise, if the regularly scheduled ping is ready before the app is refocused, the normal standby procedure is followed (e.g. the next timeout can be incremented if the connection status hasn’t changed)

### Notes for this PR
* The service worker is extended to track requests to the DHIS2 base URL and send updates to the clients based on the successes or failures of those requests
    * This is accomplished by a plugin that’s added to all the Workbox route handlers
    * Note that the SW handles ‘ping’ requests differently: it uses only the network to avoid caching, and doesn’t use the connection status plugin, since it assumes the ping is being used by the client to determine connection status (in which case the message broadcast from the SW would be redundant)
* To avoid sending a ton of messages from the SW to every client (one for every request to the DHIS2 base url it handles), the message sending function is throttled to send max 1 message a second. Messages should still be relatively frequent though in order to inform the client to delay unnecessary pings.
* Base URLs are now stored in IndexedDB on a per-app basis
    * IndexedDB is used because it is asynchronous, so the service worker can read the base URL and detect if a request is headed for the DHIS2 server or not to determine connectivity
    * Base URLs are saved on a per-app basis because it’s technically possible to configure an  app to target a DHIS2 isntance that it is not hosted on as its base URL, and after speaking with Austin, we want to support that case
    * Much of this handling happens in the App Adapter, in the ServerVersionProvider and LoginModal components
    * The @dhis2/pwa package exports some utils for performing the IndexedDB operations for base URLs

### To test locally:
_update: some of these tools are not available any more, since much of the testing has passed_
* Check out the app-runtime branch and build it
* In the PWA app’s RequestTester.js file, follow the comments to use the useDhis2ConnectionStatus hook instead of the dummy placeholder. Note that this will be imported from app-runtime when it’s released
* Run `yarn build:pwa && yarn build:adapter`
* then run `yarn temp:start` to copy the built app-runtime libs into the app-platform node_modules and start the PWA app
* The console logs and network tab can give you some cues about what’s going on under the hood

### Todos:
- [x] remove console logs after testing
- [x] remove temp scripts

[LIBS-315]: https://dhis2.atlassian.net/browse/LIBS-315?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by HendrikThePendric — COMMENTED** [review]
I've left some comments relating to code style. You have also provided me with in depth explanation about various parts of this PR on Slack and during meetings. After receiving these, I think the changes you've implemented / the approach you took here make a lot of sense.

There are still quite a few todo comments in the code and the PR is still draft, so I'll just leave this review as a "comment". I can have another look once you take this PR out of draft mode.

File: adapter/src/components/ServerVersionProvider.js [hunk 1]
```diff
@@ -1,4 +1,5 @@
 import { Provider } from '@dhis2/app-runtime'
+import { getBaseUrlByAppName } from '@dhis2/pwa/src/lib/base-url-db'
```
[the remark this answers is not in the document]
Comment (KaiVandivier) [reply adapter/src/components/ServerVersionProvider.js]: Fixed

File: pwa/src/service-worker/recording-mode.js [hunk 2]
```diff
@@ -77,19 +86,25 @@ export function shouldRequestBeRecorded({ url, event }) {
 }
 
 /** Request handler during recording mode */
-export function handleRecordedRequest({ request, event }) {
-    const recordingState = self.clientRecordingStates[event.clientId]
+export class RecordingMode extends Strategy {
+    _handle(request, handler) {
+        const { event } = handler
+        const recordingState = self.clientRecordingStates[event.clientId]
```
Comment (KaiVandivier) [inline pwa/src/service-worker/recording-mode.js:92]: This handler is refactored to a Workbox route handling strategy to take advantage of the 'plugin' functionality

File: shell/src/App.js [hunk 3]
```diff
@@ -7,9 +7,7 @@ const D2App = React.lazy(() =>
 ) // Automatic bundle splitting!
 
 const appConfig = {
-    url:
-        process.env.REACT_APP_DHIS2_BASE_URL ||
-        window.localStorage.DHIS2_BASE_URL,
+    url: process.env.REACT_APP_DHIS2_BASE_URL,
```
Comment (KaiVandivier) [inline shell/src/App.js:10]: localStorage is now checked in ServerVersionProvider.js

File: pwa/src/service-worker/dhis2-connection-status.js [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+    },
+    fetchDidSucceed: async ({ request, response }) => {
+        console.log('fetch did SUCCEED', { request, response })
+
+        if (await isRequestToDhis2Server(request)) {
+            broadcastDhis2ConnectionStatus(true)
+        }
+        return response
+    },
+}
```
Comment (KaiVandivier) [inline pwa/src/service-worker/dhis2-connection-status.js:81]: This plugin is used in `set-up-service-worker.js`

File: adapter/src/components/ServerVersionProvider.js [hunk 5]
```diff
@@ -7,43 +8,82 @@ import { LoadingMask } from './LoadingMask.js'
...
 }) => {
-    const [{ loading, error, systemInfo }, setState] = useState({
+    const [systemInfoState, setSystemInfoState] = useState({
         loading: true,
     })
+    const [baseUrlState, setBaseUrlState] = useState({
+        loading: !url,
+        error: null,
+        baseUrl: url,
+    })
```
[the remark this answers is not in the document]
Comment (KaiVandivier) [reply adapter/src/components/ServerVersionProvider.js]: Addressed in this commit: https://github.com/dhis2/app-platform/pull/718/commits/2ab16027a11d2a3e27b9f860d9b347f6b3b763d9

File: adapter/src/components/ServerVersionProvider.js [hunk 6]
```diff
@@ -7,43 +8,82 @@ import { LoadingMask } from './LoadingMask.js'
...
     useEffect(() => {
-        if (!url) {
-            setState({ loading: false, error: new Error('No url specified') })
+        // if URL prop is not set, set state to error to show login modal.
+        // Submitting valid login form with server and credentials reloads page,
+        // ostensibly with a filled url prop (now persisted locally)
+        if (!baseUrl) {
+            setBaseUrlState(state =>
+                state.loading ? state : { loading: true }
+            )
```
[the remark this answers is not in the document]
Comment (KaiVandivier) [reply adapter/src/components/ServerVersionProvider.js]: Good observations --
* using the function as the argument to setBaseUrlState and checking the condition in there allows us to read the current value of `baseUrlState.loading` without adding it as a dependency of the `useEffect` hook and avoiding potential rerender loops
* Yeah, we do want to clear the error :) I added more explicit state settings in this commit: https://github.com/dhis2/app-platform/pull/718/commits/2ab16027a11d2a3e27b9f860d9b347f6b3b763d9

**Comment by HendrikThePendric** [comment]
I had a quick look at this PR again, and while the code is looking good to me, there are still a lot of todo comments in the newly added code. Since it's not obvious to me if these are just ideas for potential future improvements, known issues that will be addressed later, or things that should actually get fixed before the PR is merged, I stopped reviewing and didn't functionally test either.

File: pwa/src/service-worker/recording-mode.js [hunk 7]
```diff
@@ -77,19 +86,25 @@ export function shouldRequestBeRecorded({ url, event }) {
...
-            return handleRecordedResponse(request, response, event.clientId)
-        })
-        .catch(error => {
-            stopRecording(error, event.clientId)
-        })
+        return handler
+            .fetch(request)
+            .then(response => {
+                return handleRecordedResponse(request, response, event.clientId)
+            })
```
[the remark this answers is not in the document]
Comment (KaiVandivier) [reply pwa/src/service-worker/recording-mode.js]: I actually prefer the explicit return here -- I feel like it communicates that something relevant is being returned from this promise.then chain (a `Response` object in this case, which the `_handle` method should resolve to), which makes it a bit easier to read

In contrast, something like `.then(response => fn(response))` could very well resolve to `undefined` if someone just wanted to save a few lines, like for `.then(response => console.log(response))`

**Comment by KaiVandivier** [comment]
Todos have been clarified

**Review by HendrikThePendric — APPROVED** [review]

**Comment by HendrikThePendric** [comment]
Just to note: I decided to approve this PR and its counterpart https://github.com/dhis2/app-runtime/pull/1203 because the code is looking good. But before merging, please take care to either remove the `console.log` statements or replace them with something that only logs in development mode, see https://github.com/dhis2/app-runtime/pull/1203#discussion_r1107385951.

**Pull request merged**